internal/services/servicefabricmanaged/service_fabric_managed_cluster_resource.go

```diff
@@ -1231,3 +1264,83 @@ func lbRulesSchema() *pluginsdk.Schema {
 		},
 	}
 }
+
+func nsRulesSchema() *pluginsdk.Schema {
+	return &pluginsdk.Schema{
+		Type:     pluginsdk.TypeList,
+		Optional: true,
+		Elem: &pluginsdk.Resource{
+			Schema: map[string]*pluginsdk.Schema{
+				"access": {
+					Type:     pluginsdk.TypeString,
+					Required: true,
+					ValidateFunc: validation.StringInSlice([]string{
+						string(managedcluster.AccessAllow),
+						string(managedcluster.AccessDeny),
+					}, false),
+				},
+				"destination_address_prefixes": {
+					Type:     pluginsdk.TypeSet,
+					Required: true,
+					Elem:     &pluginsdk.Schema{Type: pluginsdk.TypeString},
+					Set:      pluginsdk.HashString,
+				},
+				"destination_port_ranges": {
+					Type:     pluginsdk.TypeSet,
+					Required: true,
+					Elem:     &pluginsdk.Schema{Type: pluginsdk.TypeString},
+					Set:      pluginsdk.HashString,
+				},
+				"direction": {
+					Type:     pluginsdk.TypeString,
+					Required: true,
+					ValidateFunc: validation.StringInSlice([]string{
+						string(managedcluster.DirectionInbound),
+						string(managedcluster.DirectionOutbound),
+					}, false),
+				},
+				"name": {
+					Type:     pluginsdk.TypeString,
+					Required: true,
+				},
+				"priority": {
+					Type:         pluginsdk.TypeInt,
+					Required:     true,
+					ValidateFunc: validation.IntBetween(1000, 3000),
+				},
+				"protocol": {
+					Type:     pluginsdk.TypeString,
+					Required: true,
+					ValidateFunc: validation.StringInSlice([]string{
+						string(managedcluster.NsgProtocolAh),
+						string(managedcluster.NsgProtocolEsp),
+						string(managedcluster.NsgProtocolHttp),
+						string(managedcluster.NsgProtocolHttps),
+						string(managedcluster.NsgProtocolIcmp),
+						string(managedcluster.NsgProtocolTcp),
+						string(managedcluster.NsgProtocolUdp),
+					}, false),
+				},
+
+				"source_port_ranges": {
+					Type:     pluginsdk.TypeSet,
+					Required: true,
+					Elem:     &pluginsdk.Schema{Type: pluginsdk.TypeString},
+					Set:      pluginsdk.HashString,
+				},
+				"description": {
+					Type:         pluginsdk.TypeString,
+					Optional:     true,
+					ValidateFunc: validation.StringIsNotEmpty,
+				},
+
+				"source_address_prefixes": {
+					Type:     pluginsdk.TypeSet,
+					Optional: true,
+					Elem:     &pluginsdk.Schema{Type: pluginsdk.TypeString},
+					Set:      pluginsdk.HashString,
+				},
```

Comment:
can we set these to the defaults that we were setting in the code? otherwise this will be a breaking change

Reply:
```Access:                     managedcluster.AccessAllow,
				SourceAddressPrefixes:      &[]string{"0.0.0.0/0"},
				SourcePortRanges:           &[]string{"1-65535"},
				DestinationPortRanges:      &[]string{fePortStr},
				DestinationAddressPrefixes: &[]string{"0.0.0.0/0"},
				Direction:                  managedcluster.DirectionInbound,
				Name:                       fmt.Sprintf("rule%d-allow-fe", rule.FrontendPort),
				Priority:                   1000,

Reply:
@katbyte thanks for your feedback. I would like to confirm the following things:

-  Do you means that all required properties in `ns_rule` should be changed to `Optional` and set default values for them to prevent breaking change? If yes, I would like to confirm only `access`, `direction`,`priority` could be set default value in schema in my knowledge, correct? For the other properties, is there a way to set default value in schema?
- Per the [doc](https://docs.microsoft.com/en-us/azure/service-fabric/how-to-managed-cluster-networking#nsgrules), even if [networkSecurityRules](https://github.com/Azure/azure-rest-api-specs/blob/a436672b07fb1fe276c203b086b3f0e0d0c4aa24/specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/stable/2022-01-01/managedcluster.json#L1135) (a.k.a `ns_rule` in terraform) is not set in API, the priority range 3001 to 4000  network security rules are added automatically to make configurations quick and easy by defalut when creating service fabric managed clusters. However, those rules could be overridden by adding custom NSG rules ( networkSecurityRules in API a.k.a `ns_rule` in terraform) in priority range 1000 to 3000. Actually, the networkSecurityRules and [loadBalancingRules](https://github.com/Azure/azure-rest-api-specs/blob/a436672b07fb1fe276c203b086b3f0e0d0c4aa24/specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/stable/2022-01-01/managedcluster.json#L1124) are not in a one-to-one relationship as written in the previous code. And there is no default value for custom NSG rules in the API. It is optional. So I assume that terraform adds default custom NSG rules that don't follow API behavior. Exposing `ns_rule` as an optional allows users to set rules on demand instead of meaningless defaults, which is what terraform should do, WDYT?

Reply:
@sinbai - we are now allowing the user to set 
```
SourceAddressPrefixes:      &[]string{"0.0.0.0/0"},
				SourcePortRanges:           &[]string{"1-65535"},
				DestinationPortRanges:      &[]string{fePortStr},
				DestinationAddressPrefixes: &[]string{"0.0.0.0/0"},

```
- these were implied defaults, so now that we changed them to optional, thse new properties need to have those as the defaults. 
maybe even for default given users may be depending on that behaviour?

Reply:
Fixed.